02-KernelModules_CharacterDevice/src/intercom.c

```diff
@@ -0,0 +1,274 @@
+		len = msg_size;
+	status = simple_read_from_buffer(buf, len, offset, message, msg_size);
+
+	if (status < 0)
+		pr_err("Failed to read %u bytes\n", msg_size);
+	else if (status > 0) {
+		pr_info("Succefully read %d bytes\n", status);
+		msg_size -= status;
+	}
+	return status;
+}
+
+static ssize_t dev_write(struct file *fileptr, const char __user *buf,
+	size_t len, loff_t *offset)
+{
+	size_t msg_len;
+	ssize_t status;
+
+	status = string_checker(buf);
+	if (status < 0) {
+		pr_err("You should use only ASCII strings\n");
+		return status;
+	}
+
+	if (len > buf_size) {
+		pr_warn("Redusing message length from %u to %u bytes\n",
+			len, buf_size);
+		msg_len = buf_size;
+	} else
+		msg_len = len;
+	status = simple_write_to_buffer(message, msg_len, offset, buf, len);
+
+	if (status < 0)
+		pr_err("Failed to write %u bytes\n", msg_len);
+	else {
+		pr_notice("Written %u bytes\n", status);
+		msg_size = status;
+	}
+	return status;
+}
+
+static int __init intercom_init(void)
+{
+	int err;
+
+	pr_info("Intercom: Initializing kernel module\n");
+	err = alloc_chrdev_region(&chrdev, 0, 1, DEV_NAME);
+
+	if (err < 0) {
+		pr_alert("Failed to register dev major number\n");
+		return err;
+	}
+	pr_info("Major number registered correctly [%d]\n", MAJOR(chrdev));
+	intercom_cls = class_create(THIS_MODULE, CLASS_NAME);
+
+	if (IS_ERR(intercom_cls)) {
+		unregister_chrdev_region(chrdev, 1);
+		pr_alert("Failed to register device class\n");
+		return PTR_ERR(intercom_cls);
+	}
+	intercom_cls->dev_uevent = intercom_uevent;
+	err = class_create_file(intercom_cls, &sysfs_attr);
+
+	if (err < 0) {
+		pr_alert("Failed to create device class file\n");
+		return err;
+	}
+	pr_info("Device class registered correctly\n");
+
+	if (device_create(intercom_cls, NULL, chrdev, NULL, DEV_NAME) == NULL) {
+		class_destroy(intercom_cls);
+		unregister_chrdev_region(chrdev, 1);
+		pr_alert("Failed to create the device\n");
+		return -1;
+	}
+	cdev_init(&intercom_dev, &dev_fops);
+	if (cdev_add(&intercom_dev, chrdev, 1) == -1) {
+		device_destroy(intercom_cls, chrdev);
+		class_destroy(intercom_cls);
+		unregister_chrdev_region(chrdev, 1);
+		pr_alert("Failed to add the device\n");
+		return -1;
+	}
+	pr_info("Device created correctly\n");
+
+	err = create_buffer();
+	if (err)
+		goto error;
+
+	err = create_procfs();
+	if (err)
+		goto error;
+
+	return 0;
+
+error:
+	pr_err("Failed to load module\n");
+	cleanup_procfs();
+	cleanup_buffer();
+	return err;
+}
+
+static void __exit intercom_exit(void)
+{
+	cleanup_procfs();
+	cleanup_buffer();
+	device_destroy(intercom_cls, chrdev);
+	class_remove_file(intercom_cls, &sysfs_attr);
+	class_destroy(intercom_cls);
+	unregister_chrdev_region(chrdev, 1);
+	pr_info("Exiting intercom\n");
+}
+
+module_init(intercom_init);
+module_exit(intercom_exit);
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Nikita Kvochka <yorkened@gmail.com>");
+MODULE_DESCRIPTION("Simple char dev for text messaging between users");
+MODULE_VERSION("0.1");
```

Comment:
add empty line at the end of file
```suggestion
MODULE_VERSION("0.1");

```